algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala

```diff
@@ -0,0 +1,342 @@
+ * @param time current timestamp of this instance.
+ *             Used with conf.windowSize to expire buckets.
+ */
+case class ExpHist(conf: ExpHist.Config, buckets: Vector[ExpHist.Bucket], total: Long, time: Long) {
+  import ExpHist.Bucket
+
+  /**
+   * Steps this instance forward to the new supplied time. Any
+   * buckets with a timestamp <= (newTime - conf.windowSize) will be
+   * evicted.
+   *
+   * @param newTime the new current time.
+   * @return ExpHist instance stepped forward to newTime.
+   */
+  def step(newTime: Long): ExpHist =
+    if (newTime <= time) this
+    else {
+      val (dropped, filtered) = conf.dropExpired(buckets, newTime)
+      copy(time = newTime, buckets = filtered, total = total - dropped)
+    }
+
+  /**
+   * Increment ExpHist by 1 at the supplied timestamp.
+   */
+  def inc(timestamp: Long): ExpHist = add(1L, timestamp)
+
+  /**
+   * Increment ExpHist by delta at the supplied timestamp.
+   */
+  def add(delta: Long, timestamp: Long): ExpHist = {
+    val self = step(timestamp)
+    if (delta == 0) self
+    else self.addAllWithoutStep(Vector(Bucket(delta, timestamp)), delta)
+  }
+
+  /**
+   * Efficiently add many buckets at once.
+   *
+   * @param unsorted [bucket]. All timestamps must be >= this.time.
+   * @return ExpHist instance with all buckets added, stepped
+   *         forward to the most timestamp in `unsorted`.
+   */
+  def addAll(unsorted: Vector[Bucket]): ExpHist =
+    if (unsorted.isEmpty) this
+    else {
+      val sorted = unsorted.sorted(Ordering[Bucket].reverse)
+      val delta = sorted.map(_.size).sum
+      val timestamp = sorted.head.timestamp
+      if (delta == 0)
+        step(timestamp)
+      else {
+        addAllWithoutStep(sorted, delta).step(timestamp)
+      }
+    }
+
+  // This internal method assumes that the instance is stepped forward
+  // already, and does NOT try to step internally. It also assumes
+  // that `items` is sorted in ASCENDING order, with newer items on
+  // the right side, and that `items.map(_.size).sum == delta`.
+  private[ExpHist] def addAllWithoutStep(items: Vector[Bucket], delta: Long): ExpHist = {
+    val inputs = items ++ buckets
+    val desiredBuckets = Canonical.bucketsFromLong(total + delta, conf.l)
+    copy(
+      buckets = ExpHist.rebucket(inputs, desiredBuckets),
+      total = total + delta)
+  }
+
+  def oldestBucketSize: Long = if (total == 0) 0L else buckets.last.size
+
+  /**
+   * Smallest possible count seen in the last conf.windowSize
+   * timestamps.
+   */
+  def lowerBoundSum: Long = total - oldestBucketSize
+
+  /**
+   * Largest possible count seen in the last conf.windowSize
+   * timestamps.
+   */
+  def upperBoundSum: Long = total
+
+  /**
+   * Estimate of the count seen across the last conf.windowSize
+   * timestamps. Guaranteed to be within conf.epsilon of the true
+   * count.
+   */
+  def guess: Double =
+    if (total == 0) 0.0
+    else (total - (oldestBucketSize - 1) / 2.0)
+
```

Comment:
don't we want `approximate[Double]`, since we want the estimate to fall in the middle of the range?